File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/cats/DeepBind.scala
package arrow.benchmarks.effects.scala.cats

import cats.effect.IO

object DeepBind {

  def fib(n: Int): IO[BigInt] =
    if (n <= 1) IO(n)
    else
      fib(n - 1).flatMap { a =>
        fib(n - 2).flatMap(b => IO(a + b))
      }
}

File: modules/benchmarks/arrow-kio-benchmarks/src/main/kotlin/kio/DeepBind.kt
package arrow.benchmarks.effects.kio

import it.msec.kio.UIO
import it.msec.kio.effect
import it.msec.kio.flatMap
import it.msec.kio.runtime.Runtime

object DeepBind {

  fun loop(n: Int): UIO<Int> =
    if (n <= 1) effect { n }
    else loop(n - 1).flatMap { a ->
      loop(n - 2).flatMap { b -> effect { a + b } }
    }

  fun fib(depth: Int) = Runtime.unsafeRunSync(loop(depth))
}
